[marshal methods] Make AndroidEnvironmentInternal.UnhandledException …

grendello opened the pull request:
…public

Fixes: https://github.com/dotnet/android/issues/10602
Context: https://github.com/dotnet/android/commit/8bc7a3e84f95e70fe12790ac31ecd97957771cb2

This is in the category of "how the hell has it worked before?", as every time the `AndroidEnvironmentInternal.UnhandledException` method living in `Mono.Android.Runtime.dll` would be called, the runtime would respond with a method not accessible exception, rightly so as both the type and the method were internal and visible only to `Mono.Android.dll`.

I guess we haven't had many unhandled exceptions in marshal method wrappers over the years?

This PR fixes the problem by making both the type and the method public. This is the simplest way to fix the issue.

grendello committed:
[marshal methods] Make AndroidEnvironmentInternal.UnhandledException public

Fixes: https://github.com/dotnet/android/issues/10602
Context: https://github.com/dotnet/android/commit/8bc7a3e84f95e70fe12790ac31ecd97957771cb2

This is in the category of "how the hell has it worked before?", as every time the
`AndroidEnvironmentInternal.UnhandledException` method living in `Mono.Android.Runtime.dll`
would be called, the runtime would respond with a method not accessible exception, rightly
so as both the type and the method were internal and visible only to `Mono.Android.dll`.

I guess we haven't had many unhandled exceptions in marshal method wrappers over the years?

This PR fixes the problem by making both the type and the method public. This is the simplest
way to fix the issue.

jonathanpeppers reviewed (commented):
Can we write some Mono.Cecil code to make this public at build time? So, no one can use it at C# compile time?

This would happen in the marshal methods MSBuild steps, if they run.

grendello committed:
Rewrite Mono.Android.Runtime.dll if marshal methods are enabled

The rewrite changes visibility of the `Android.Runtime.AndroidEnvironmentInternal` type
from `internal` to `public` after all the linking and marshal method processing is done.

grendello committed:
Add a test

grendello committed:
Address feedback

grendello commented:
> @grendello today I learned about `[assembly: IgnoresAccessChecksToAttribute ("Mono.Android")]`:
> 
>     * https://github.com/search?q=repo%3Adotnet%2Fruntime+IgnoresAccessChecksToAttribute&type=code
> 
> 
> This appears to be a runtime feature, that Mono even implements? I can't find docs on it, though.

Yeah, it appears to be undocumented which has me worried a bit. Mono does have support for it in https://github.com/dotnet/runtime/blob/0bbe2a937db7b916ca94849de3a3d16f23f8b17f/src/mono/mono/metadata/assembly.c#L1707-L1710

The comment there appears to suggest it might be meant only for dynamically generated modules? 

> 
> Can we simply stamp this attribute on every assembly? It could probably be done in the existing MSBuild tasks that write the marshal methods?

TBH, there's too much uncertainty about this attribute and its role. I wouldn't feel comfortable using it. It also appears to be a nuclear solution to the problem - it would essentially let anyone access anything in our assemblies, thus potentially creating maintenance problems for the future (people could start relying on internals being there because of that attribute)

grendello commented on src/Mono.Android.Runtime/Android.Runtime/AndroidEnvironmentInternal.cs:
```diff
@@ -6,7 +6,7 @@ internal static class AndroidEnvironmentInternal
 		internal static Action<Exception>? UnhandledExceptionHandler;
 
-		internal static void UnhandledException (Exception e)
+		public static void UnhandledException (Exception e)
```
It's an API and ABI change, we can't do it in a stable project and the fix needs to be backported to net10.
